apis/apiextensions/v1beta1/composition_types.go

```diff
@@ -217,10 +221,16 @@ type Transform struct {
 	// +optional
 	Math *MathTransform `json:"math,omitempty"`
 
-	// Map uses the input as a key in the given map and returns the value.
+	// Map uses the input as a key in the given map whose values can only be
+	// strings and returns the value.
 	// +optional
 	Map *MapTransform `json:"map,omitempty"`
 
+	// BoolMap uses the input as a key in the given map whose values can only
+	// be booleans and returns the value.
+	// +optional
+	BoolMap *MapTransform `json:"boolMap,omitempty"`
```

Comment:
I appreciate the eye for backwards compatibility here, but do you think we could just adapt the existing implementation to infer a boolean if value in `map[string]string` is `true` or `false`? I believe this could be done without breaking API compatability

Reply:
I had a similar thought. Maybe we could add a flag that controlled that behaviour? I'm rusty on how this API works, but I was thinking something like:

```yaml
transforms:
- type: map
  # When this is set the strings 'true' and 'false' are
  # interpreted as boolean true/false.
  inferBooleanValues: true
  map:
    key1: true
```

This would maintain API and behaviour compatibility for existing folks without needing to create a new kind of map.

Reply:
I'm thinking the cases where we'd expand this to other types like `small: 1024, big: 4096` where the value is integer. In that case, we'd have a separate `inferIntegerValues`. Maybe we can have one field named `valueType: [int, bool, string, []int, ....]` (possibly use an enum for possible types or use reflect package for arbitrary input) and we'd cast the values into that type. We could have only one field and allow users to tell the type with `string` being implicit default to keep the existing behavior. Though we can also just say it wouldn't go much further than `string, int, bool` types anyway, so two additional fields would be good enough. What do you think?

Reply:
I like the idea of `valueType` 👍

Reply:
Yeah, that sounds like a more future proof approach.

Reply:
Actually it seems like it's harder to get an option field for MapTransform. In order to do the following:
```yaml
transforms:
- type: map
  anOptionField: aValue
  map:
    key1: true
```

We'd have to add that option to the same level as transforms [here](https://github.com/crossplane/crossplane/blob/e2d7278/apis/apiextensions/v1beta1/composition_types.go#L210). And I wanted to cheated by adding it to under `map` like the following:
```
transforms:
- type: map
  map:
    anOptionField: aValue # maybe api-server reserves that "anOptionField" keyword for us
    key1: true
```

But we get the following error as expected during the registration of `Composition` CRD:
```
Error: failed to install CRD crds/apiextensions.crossplane.io_compositions.yaml: CustomResourceDefinition.apiextensions.k8s.io "compositions.apiextensions.crossplane.io" is invalid: spec.versions[1].schema.openAPIV3Schema.properties[spec].properties[resources].items.properties[patches].items.properties[transforms].items.properties[map].additionalProperties: Forbidden: additionalProperties and properties are mutual exclusive
```

But when you think about it, maybe this `valueType` could actually go for any transform? So, we'd do the following:
```yaml
- type: map
  valueType: bool
  map:
    key1: true
- type: map
  valueType: float64
  math:
    multiply: "1.2"
```

And then there wouldn't be a change in `map` transform specifically; we'd cast any value returned from the chosen transform resolver into the given `valueType`. Is everyone cool about this?

Reply:
We have discussed this a bit more offline and @negz brought up a great idea. We'd just add a new transform type to do a type conversion so that output from any transform could be input to that conversion transform. I'll update the PR to do that.